src/server/components/zitadel/utils.ts

```diff
@@ -0,0 +1,183 @@
+        }
+
+        const hrStart = process.hrtime();
+
+        const axiosInstance = axios.create();
+
+        const response = await axiosInstance({
+            method: 'post',
+            url: `${zitadelUri}/oauth/v2/introspect`,
+            auth: {
+                username: clientId,
+                password: clientSecret,
+            },
+            'axios-retry': {
+                retries: 3,
+            },
+            params: {
+                token,
+            },
+        });
+
+        const {active} = response.data;
+        const result = Boolean(active);
+
+        ctx.log(`Token introspected successfully within: ${getDuration(hrStart)} ms`);
+        return result;
+    } catch (e) {
+        ctx.logError('Failed to introspect token', e);
+        return false;
+    }
+};
+
+export const refreshTokens = async (ctx: AppContext, refreshToken?: string) => {
+    ctx.log('Refreshing tokens');
+
+    if (!clientId) {
+        throw new Error('Missing CLIENT_ID in env');
+    }
+    if (!clientSecret) {
+        throw new Error('Missing CLIENT_SECRET in env');
+    }
+    if (!refreshToken) {
+        throw new Error('Token not provided');
+    }
+
+    try {
+        const axiosInstance = axios.create();
+
+        const response = await axiosInstance({
+            method: 'post',
+            url: `${zitadelUri}/oauth/v2/token`,
+            auth: {
+                username: clientId,
+                password: clientSecret,
+            },
+            params: {
+                refresh_token: refreshToken,
+                grant_type: 'refresh_token',
+                scope: 'openid profile',
+            },
+        });
+
+        ctx.log('Tokens refreshed successfully');
+
+        return {accessToken: response.data.access_token, refreshToken: response.data.refresh_token};
+    } catch (e) {
+        ctx.logError('Failed to refresh tokens', e);
+        return {accessToken: undefined, refreshToken: undefined};
+    }
+};
+
+export const fetchServiceUserAccessToken = async (ctx: AppContext) => {
+    if (!zitadelUri) {
+        throw new Error('Missing ZITADEL_URI in env');
+    }
+    if (!serviceClientId) {
+        throw new Error('Missing SERVICE_CLIENT_ID in env');
+    }
+    if (!serviceClientSecret) {
+        throw new Error('Missing SERVICE_CLIENT_SECRET in env');
+    }
+    if (!zitadelProjectId) {
+        throw new Error('Missing ZITADEL_PROJECT_ID in env');
+    }
+
+    try {
+        ctx.log('Fetching service user access token');
+
+        const response = await axios.post(
+            `${zitadelUri}/oauth/v2/token`,
+            new URLSearchParams({
+                grant_type: 'client_credentials',
+                scope: `openid profile urn:zitadel:iam:org:project:id:${zitadelProjectId}:aud`,
+            }),
+            {
+                auth: {
+                    username: serviceClientId,
+                    password: serviceClientSecret,
+                },
```

Comment:
'axios-retry'?

or we can create one axios instance for this file:
```
const axiosInstance = axios.create();
axiosRetry(axiosInstance, { retries: 3 });
```

Reply:
done